lib/stack_nav.py

```diff
@@ -0,0 +1,63 @@
+"""A stack-based naviation system.
+
+Maintains a stack of states through which a user can navigate.
+Think Android's back button model.
+
+Every screen is modeled as a function that takes one argument "state".
+It can modify state and should either return
+* None (which means go back one level)
+* A new function (which adds to the stack)
+* A keyword like NEXT_EXIT (exit nav) or NEXT_TOP (go to top of stack)
+
+Children get a a shallow clone of the current state, going back will
+lead to a screen with the original state.
+"""
+
+___dependencies___ = ["dialogs", "database"]
+___license___ = "MIT"
+
+import dialogs, database
+
+NEXT_EXIT = "exit" # Leave navigation
+NEXT_TOP = "init" # Go to top of stack
+def nav(init_fn, init_state={}):
+    stack = [(init_fn, init_state)]
+    while len(stack):
```

Comment:
empty arrays evaluate as `False`, so `len` is not required. The following should be enough:

```python
while stack:
    # ...
```